iree/compiler/Conversion/LinalgToLLVM/LinalgTileAndVectorizePass.cpp

```diff
@@ -0,0 +1,161 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "iree/compiler/Conversion/CodegenUtils/MarkerUtils.h"
+#include "iree/compiler/Conversion/CodegenUtils/MatmulCodegenStrategy.h"
+#include "iree/compiler/Conversion/LinalgToLLVM/KernelDispatch.h"
+#include "mlir/Conversion/StandardToSPIRV/ConvertStandardToSPIRV.h"
+#include "mlir/Dialect/Linalg/Transforms/Hoisting.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+#define DEBUG_TYPE "iree-linalg-to-llvm-tile-and-vectorize"
+
+namespace mlir {
+namespace iree_compiler {
+
+namespace {
+template <typename LinalgOpTy>
+struct TileWorkgroups : public linalg::LinalgBaseTilingPattern {
+  using Base = linalg::LinalgBaseTilingPattern;
+  TileWorkgroups(MLIRContext *context, linalg::LinalgTilingOptions options,
+                 linalg::LinalgMarker marker, PatternBenefit benefit = 1)
+      : Base(LinalgOpTy::getOperationName(), context, options, marker,
+             benefit) {}
+  LogicalResult matchAndRewrite(Operation *op,
+                                PatternRewriter &rewriter) const override {
+    // Find the parent FuncOp before tiling. If tiling succeeds, the op will be
```

Comment:
Comment seems off. The `LinalgBaseTilingPattern` doesnt erase the op (and you are erasing it here as you should). Also doesnt seem like you need `funcOp` at all.

Reply:
Done!